### PR TITLE
[nan-842] update refresh logic based on spec change

### PR DESCRIPTION
**Pull request by khaliqgant**
## Describe your changes
Updated specs for the refresh token logic. 1 day and time cap on 23 hour limit on the cron.

## Issue ticket number and link
NAN-842

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

**Review by bodinsamuel — APPROVED** [review]

File: packages/server/lib/refreshTokens.ts [hunk 1]
```diff
@@ -28,15 +28,23 @@ export function refreshTokens(): void {
 }
 
 export async function exec(): Promise<void> {
+    const startTime = Date.now();
+    const maxDuration = 23 * 60 * 60 * 1000;
+
     logger.info('[refreshTokens] starting');
 
-    const staleConnections = await connectionService.getOldConnections({ days: 7, limit: 500 });
+    const staleConnections = await connectionService.getOldConnections({ days: 1 });
```
Comment (TBonnin) [inline packages/server/lib/refreshTokens.ts]: can we still have a limit (even if very high)? Just to make sure any refresh token problem don't cascade into making this call returns a huge number of connections

**Review by TBonnin — APPROVED** [review]
would it make sense to run the cron more often (instance of just once a day). Since it is running once a day, a token be stale for almost 48h. Also it will help catching problem way more rapidely.

File: packages/server/lib/refreshTokens.ts (hunk 1 above)
Comment (khaliqgant) [reply packages/server/lib/refreshTokens.ts]: Point taken about not returning a huge number of connections. I think I'd rather chunk this actually, will update this tomorrow. Thanks

**Comment by khaliqgant** [comment]
Made some adjustments to use a lock, run every 45 minutes with a limit of 250 connections at a time. Removed time limit since we are now using a lock. This lock also prevents multiple instances of the cron running on the server @bodinsamuel @TBonnin

File: packages/server/lib/refreshTokens.ts [hunk 2]
```diff
@@ -30,28 +30,39 @@ export function refreshTokens(): void {
...
 
-    const staleConnections = await connectionService.getOldConnections({ days: 7, limit: 500 });
+    const today = new Date();
+    const lockKey = parseInt(`${today.getFullYear()}${today.getMonth() + 1}${today.getDate()}`);
 
-    logger.info(`[refreshTokens] found ${staleConnections.length} stale connections`);
+    // Lock to prevent multiple instances of this cron job from running at the same time
+    await db.knex.transaction(async (trx) => {
+        const { rows } = await trx.raw<{ rows: { pg_try_advisory_xact_lock: boolean }[] }>(`SELECT pg_try_advisory_xact_lock(?);`, [lockKey]);
+        if (!rows || rows.length <= 0 || !rows[0]!.pg_try_advisory_xact_lock) {
```
Comment (TBonnin) [inline packages/server/lib/refreshTokens.ts]: can probably be simplify by using optional chaining `if(!rows?.[0]?.pg_try_advisory_xact_lock)

File: packages/server/lib/refreshTokens.ts [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
+import { db, errorManager, ErrorSourceEnum, connectionService } from '@nangohq/shared';
 import { stringifyError, getLogger, metrics } from '@nangohq/utils';
 import tracer from 'dd-trace';
 import { logContextGetter } from '@nangohq/logs';
 
 const logger = getLogger('Server');
 
 export function refreshTokens(): void {
-    cron.schedule('0 6 * * *', () => {
+    cron.schedule('*/45 * * * *', async () => {
```
Comment (TBonnin) [inline packages/server/lib/refreshTokens.ts:10]: I am curious to know why you picked 45 mins

File: packages/server/lib/refreshTokens.ts [hunk 4]
```diff
@@ -30,28 +30,39 @@ export function refreshTokens(): void {
...
-            logger.error(`[refreshTokens] connectionId: ${connection_id} is missing account_id`);
-            continue;
-        }
+        for (const staleConnection of staleConnections) {
+            const { connection_id, environment_id, provider_config_key, account_id } = staleConnection;
+
+            if (typeof account_id !== 'number') {
+                logger.error(`[refreshTokens] connectionId: ${connection_id} is missing account_id`);
+                continue;
+            }
```
Comment (TBonnin) [inline packages/server/lib/refreshTokens.ts]: How can a connection not have an account id?

File: packages/shared/lib/services/connection.service.ts [hunk 5]
```diff
@@ -446,19 +446,24 @@ class ConnectionService {
         return result;
     }
 
-    public async getOldConnections({ days, limit }: { days: number; limit: number }): Promise<NangoConnection[]> {
+    public async getOldConnections({ days, limit }: { days: number; limit?: number }): Promise<NangoConnection[]> {
```
Comment (TBonnin) [inline packages/shared/lib/services/connection.service.ts]: why optional?

**Review by TBonnin — APPROVED** [review]

File: packages/server/lib/refreshTokens.ts (hunk 4 above)
Comment (khaliqgant) [reply packages/server/lib/refreshTokens.ts]: I was trying to use the `NangoConnection` type which had `account_id` as potentially undefined. I updated the method to have an account_id that is set and remove this check with https://github.com/NangoHQ/nango/pull/2105/commits/8596099ad95ab9925fc7e81af0b9731d99741475

File: packages/shared/lib/services/connection.service.ts (hunk 5 above)
Comment (khaliqgant) [reply packages/shared/lib/services/connection.service.ts]: Left over from the previous iteration. Updated with https://github.com/NangoHQ/nango/pull/2105/commits/8596099ad95ab9925fc7e81af0b9731d99741475

File: packages/server/lib/refreshTokens.ts (hunk 3 above)
Comment (khaliqgant) [reply packages/server/lib/refreshTokens.ts:10]: Arbitrarily. But it was a good middle ground between 30 and 1 hour. With 45 minutes it'll take about 2 days to go through the connections. I want to see how long an iteration takes and can tweak it further if needed.

File: packages/server/lib/refreshTokens.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/server/lib/refreshTokens.ts]: Updated

**Pull request merged**